File: 4. Median of Two Sorted Arrays/Median of Two Sorted Arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
        int ans[n + m];
        double median;
        int i = 0, j = 0, index = 0;
        while(i < m && j < n){
            if(nums1[i] <= nums2[j]){
                ans[index++] = nums1[i++];
            } else {
                ans[index++] = nums2[j++];
            }
        }
        while(i < m) ans[index++] = nums1[i++];
        while(j < n) ans[index++] = nums2[j++];
        if(index % 2){
            median = (double)ans[index / 2];
        } else {
            median = (ans[index / 2] + ans[index / 2 - 1]) / 2.0;
        }
        return median;
    }
};

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
        int total = m + n;
        //如果总长度是奇数
        if(total & 0x1){
            return (double)findKth(nums1, 0, nums2, 0, total / 2 + 1);
        } else {
           // cout << findKth(nums1, 0, nums2, 0, total / 2) << endl;
            //cout << findKth(nums1, 0, nums2, 0, total / 2 + 1) << endl;
            
            return (findKth(nums1, 0, nums2, 0, total / 2) + findKth(nums1, 0, nums2, 0, total / 2 + 1)) / 2.0;
        }
    }
    //寻找A和B数组中第k小的数
    //i,j分别为指针 指向开始查找的位置
    int findKth(vector<int>& A, int i, vector<int>& B, int j, int k){
        //i超过A数组的边界 说明A中没有要查找的元素 直接返回B中的元素
        if(i >= A.size()) return B[j + k - 1];
        //j超过B数组的边界 说明B中没有要查找的元素 直接返回A中的元素
        if(j >= B.size()) return A[i + k - 1];
        if(k == 1) return min(A[i], B[j]);
        //两个数组同时二分
        //每个数组找到第k/2个元素 如果越界 取最大值INT_MAX
        int midA = ((i + k / 2 - 1) < A.size()) ? A[i + k / 2 - 1] : INT_MAX;
        int midB = ((j + k / 2 - 1) < B.size()) ? B[j + k / 2 - 1] : INT_MAX;
        if(midA < midB){
            return findKth(A, i + k / 2, B, j, k - k / 2);
        } else {
            return findKth(A, i, B, j + k / 2, k - k / 2);
        }
        
    }
       
};
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
        if(m > n) return findMedianSortedArrays(nums2, nums1);
        int lo = 0;
        int hi = m;
        while(lo <= hi){
            int cut1 = (lo + hi) / 2;
            int cut2 = (m + n) / 2 - cut1;
            int l1 = (cut1 == 0) ? INT_MIN : (nums1[cut1 - 1]);
            int l2 = (cut2 == 0) ? INT_MIN : (nums2[cut2 - 1]);
            int r1 = (cut1 == m) ? INT_MAX : (nums1[cut1]);
            int r2 = (cut2 == n) ? INT_MAX : (nums2[cut2]);
            if(l1 > r2){
                hi = cut1 - 1;
            } else if(l2 > r1) {
                lo = cut1 + 1;
            } else {
                return (m + n) % 2 ? min(r1, r2) : (max(l1, l2) + min(r1, r2)) / 2.0;
            }
        }
        return 0.0;
    }
};

File: 4. Median of Two Sorted Arrays/Median of Two Sorted Arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        m = len(nums1)
        n = len(nums2)
        ans = [0] * (n + m)
    
        i, j , index = 0, 0, 0
        while i < m and j < n:
            if nums1[i] <= nums2[j]:
                ans[index] = nums1[i]
                i = i + 1
                index = index + 1
            else:
                ans[index] = nums2[j]
                j = j + 1
                index = index + 1
            
        
        while i < m: 
            ans[index] = nums1[i]
            i = i + 1
            index = index + 1
        while j < n:
            ans[index] = nums2[j]
            j = j + 1
            index = index + 1
       # print ans
        if index % 2:
            median = ans[index / 2]
        else:
            median = (ans[index / 2] + ans[index / 2 - 1]) / 2.0
        
        return median;
class Solution(object):
    
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        
        def findKth(A, i, B, j, k):
        
            if i >= len(A): return B[j + k - 1]

            if j >= len(B): return A[i + k - 1]
            if k == 1: return min(A[i], B[j])

            midA = A[i + k / 2 - 1] if (i + k / 2 - 1) < len(A) else sys.maxsize
            midB = B[j + k / 2 - 1] if (j + k / 2 - 1) < len(B) else sys.maxsize

            if midA < midB:
                return findKth(A, i + k / 2, B, j, k - k / 2)
            else:
                return findKth(A, i, B, j + k / 2, k - k / 2)
        m = len(nums1)
        n = len(nums2)
        
        total = m + n
        
        if total % 2 == 1:
            return findKth(nums1, 0, nums2, 0, total // 2 + 1)
        else:
            return (findKth(nums1, 0, nums2, 0, total // 2) + findKth(nums1, 0, nums2, 0, total // 2 + 1)) / 2.0
            
class Solution(object):
    
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        
        m , n = len(nums1), len(nums2)
        if m > n: 
            nums1, m, nums2, n = nums2, n, nums1, m
        lo = 0
        hi = m
        while lo <= hi:
            cut1 = (lo + hi) >> 1
           # print cut1
            cut2 = ((m + n) >> 1) - cut1
           # print cut2
            l1 = -sys.maxsize - 1 if (cut1 == 0) else (nums1[cut1 - 1])
            l2 = -sys.maxsize - 1 if (cut2 == 0) else (nums2[cut2 - 1])
            r1 = sys.maxsize if (cut1 == m) else (nums1[cut1])
            r2 = sys.maxsize if (cut2 == n) else (nums2[cut2])
            if l1 > r2:
                hi = cut1 - 1
            elif l2 > r1:
                lo = cut1 + 1
            else:
                if (m + n) % 2:
                    return min(r1, r2)
                else:
                    return (max(l1, l2) + min(r1, r2)) / 2.0
                
            
        
        return 0.0;
